### PR TITLE
[WIP] [new] Screenshots and screen recordings of tests 

### PR DESCRIPTION
**Pull request by noomorph**
The pull request aims to add the following features:

1. Automatic screen recording on iOS/Android with two configurable options: `--artifacts-location [rootDir]` (where videos will be saved) and `--record-videos [all|failing|none]` behavior (`none` is default and means that recording is *disabled*).
2. Automatic screenshots (beforeEach and afterEach) on iOS/Android with two configurable options: `--artifacts-location [rootDir]` (where screenshots will be saved) and `--take-screenshots [all|failing|none]` behavior (`none` is default and means that screenshotting is *disabled*). 
3. Automatic log recording during test on iOS/Android with two configurable options: `--artifacts-location [rootDir]` (where logs will be saved) and `--record-logs [all|failing|none]` behavior (`none` is default and means that recording is *disabled*).
4. Out-of-box support for [mocha] and [jest] test runners:
* `detox init -r mocha` and `detox init -r jest` should scaffold configuration files which are sufficient for the new artifact recording features and (hopefully) flexible enough not to require manual updates on the user side.
* the latter is being solved by introducing `detox<->mocha` and `detox<->jest` adapters which encapsulate the tricky parts to enable executing long-running hooks (e.g. `before (beforeAll)`, `beforeEach`, `afterEach`, `after (afterAll)` in `mocha` and `jest`) together with ability to receive current running test status and name (not trivial thing in `jest`). That's a thing strongly required by the artifacts feature. 
5. Artifacts feature should be fairly decoupled from detox internals - decoupled enough to grant a possibility to extract it to a separate plugin with ease in more distant future (if we'll decide we need it).

### Current status

The PR’s author (@noomorph) has returned from the vacation. He is working on documentation this week, stabilizing code, improving error messages and unit tests also.

### TODO list

* [x] make unit tests pass again
* [x] complete and fix Android video recording
* [x] complete and fix taking screenshots on Android
* [x] complete and fix taking screenshots on iOS
* [x] complete and fix iOS video recording
* [x] pass initial pull request review
* [x] rewrite iOS logger to follow `.start().stop().save().discard()` interface
* [x] write `adb logcat`-based logger implementation with the aforementioned interface
* [x] rewrite to make adding new type of artifacts even easier
* [x] synchronous .kill() interface when users press Ctrl+C (`SIGINT` or other sort of emergency exit).
* [x] improve adb logcat logger - keep recording logs even after device relaunch (when `pid` changes)
* [x] simple exception handling in `onStart`, `onBeforeTest`, ... hooks
* [x] revert .kill() interface idea in favor of state machine in base `ArtifactRecording.js` class
* [x] fix `unknown` artifact plugin name in `onIdleCallback` error 
* [x] fix `upload_artifact.sh` script (from previous PR) error (`aws: command not found`) on Travis CI
* [x] fix bug with first screenshot on iOS
* [ ] complete unit tests for core functionality

Documentation:
* [x] update [Getting Started](https://github.com/wix/detox/blob/feature/test-artifacts/docs/Introduction.GettingStarted.md).
* [x] update [Migration Guide](https://github.com/wix/detox/blob/feature/test-artifacts/docs/Guide.Migration.md).
* [x] update [test lifecycle](https://github.com/wix/detox/blob/feature/test-artifacts/docs/APIRef.TestLifecycle.md) about `beforeEach` and `afterEach`
* [x] update [detox object](https://github.com/wix/detox/blob/feature/test-artifacts/docs/APIRef.DetoxObjectAPI.md) about `detox.beforeEach(testSummary)` and `detox.afterEach(testSummary)`.
* [x] update [Jest Guide](https://github.com/wix/detox/blob/feature/test-artifacts/docs/Guide.Jest.md) about `detox init -r jest`.
* [x] update [detox-cli](https://github.com/wix/detox/blob/feature/test-artifacts/docs/APIRef.DetoxCLI.md) with new artifact flags and breaking change that `--record-logs all` should be specified explicitly in order to retain the previous behavior.
* [x] update [Enabling artifacts](https://github.com/wix/detox/blob/feature/test-artifacts/docs/APIRef.Artifacts.md) section with first steps and troubleshooting
* [x] add a pitfalls subsection: e.g. `Ctrl+C does not work correctly with Jest`, `Metal is not supported on VM`, or maybe `Increased flakiness with artifacts recording`, and so on...

Nice to haves:
* extend configuration for screenshot plugin - add an ability to specify at which phase you want screenshots to be taken: beforeEach, afterEach or both. E.g., `all:beforeEach,afterEach` means that screenshots will be taken in every test two times: in a global beforeEach and in a global afterEach. One more example, `failing:afterEach` will mean that screenshots will be taken only on failing tests in a global afterEach. In future we might add `beforeTest` and `afterTest` phases which will surround `test()` or `it()` statement.
* integration tests for artifact plugins
* merge from branch e2e suite for`detox/src/runners/jest/DetoxJestAdapter.js`
* merge from branch e2e suite for `detox/src/runners/mocha/DetoxMochaAdapter.js`
* improve logging in verbose mode (e.g. add prefix by process name and pid to `exec` logs)
* add extensive logging for artifacts plugins
* specific exception handling (e.g. Metal is not supported on VM)

Links:

* the former pull request: https://github.com/wix/detox/pull/541 .
* the parent issue: https://github.com/wix/detox/issues/171

Hopefully, I'll be moving quite-quite fast.
Stay tuned.

### CODE REVIEW

**Comment by noomorph** [comment]
I thought it might be curious to share a video of `detox/test` project run (from yesterday) with android video recording turned on.

https://youtu.be/mLVw3jowr4k
<a href="https://www.youtube.com/embed/mLVw3jowr4k">
click to see video
</a>

A nice improvement (since the former PR) is that Detox copies recorded videos in parallel. Namely, we don't wait for all tests to complete before we start pulling videos from the device.

P. S. On the other hand, I have a silly concern whether framerates are apt to drop (...or maybe not?) when `adb screenrecord` and `adb pull` run concurrently. If this proves to be an issue, I might revert the behavior to the former one or alternatively, add an extra flag, but, in any case, this "gut feeling" needs to be either proved or disproved.

**Comment by noomorph** [comment]
@rotemmiz , I made a (relatively) minor change to artifacts API.

`record('artifact path').start().stop().save()` became `record().start().stop().save('artifact path')`.

It gives us a superpower to give different names to artifacts depending on test status (`passed|failed`). Hope you'll like it: `✔ 1. Passing test` and `✗ 2. Failing test`. 😛 

This becomes very handy with option *--record-whatever* ***all***.

screen shot 2018-05-18 at 19 42 35

**Comment by noomorph** [comment]
🎉 I've got iOS simulator logging working, hooray!

**Comment by noomorph** [comment]
@rotemmiz, check out the verbosity level I did for Android logs:

[test.log](https://github.com/wix/detox/files/2023173/test.log)

File: detox/src/runners/jest/DetoxJestAdapter.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      await Promise.resolve().then(t.shift()).catch(()=>{});
+    }
+  }
+
+  specStarted(result) {
+    if (result.pendingReason) {
+      return;
+    }
+
+    const spec = Object.freeze({
```
Comment (rotemmiz) [inline detox/src/runners/jest/DetoxJestAdapter.js]: This object is going inside Detox implementation, which just reads from it (correct me if I'm wrong please)
Why do we need to freeze it ?

Comment (rotemmiz) [reply detox/src/runners/jest/DetoxJestAdapter.js]: I suggest we get rid of all `Object.freeze`, especially if the consumers are us.

**Comment by noomorph** [comment]
Quick update. I've removed artifacts folders' numbering to avoid potential issues with parallel e2e test execution on multiple devices.

Now the folders look this way:
 
screenshot from 2018-05-21 19-00-29

File: detox/src/devices/android/ADB.js [hunk 2]
```diff
@@ -96,13 +126,86 @@ class ADB {
...
+      height: parseInt(height, 10)
+    };
+  }
+
+  async getFileSize(deviceId, path) {
+    const {stdout} = await this.adbCmd(deviceId, `shell wc -c ${path}`);
+    return parseInt(stdout, 10);
+  }
+
+  screencap(deviceId, path) {
```
Comment (rotemmiz) [inline detox/src/devices/android/ADB.js]: is this supposed to be an async function ?

File: detox/src/devices/android/ADB.js [hunk 3]
```diff
@@ -96,13 +126,86 @@ class ADB {
...
+
+    if (time) {
+      logcatArgs.push('-T');
+      logcatArgs.push(time);
+    }
+
+    return this.spawn(deviceId, ['logcat', ...logcatArgs]);
+  }
+
+  pull(deviceId, src, dst = '') {
```
Comment (rotemmiz) [inline detox/src/devices/android/ADB.js]: async ?

File: detox/src/devices/android/ADB.js [hunk 4]
```diff
@@ -96,13 +126,86 @@ class ADB {
...
+    return this.adbCmd(deviceId, `pull "${src}" "${dst}"`);
+  }
+
+  rm(deviceId, path, force = false) {
```
Comment (rotemmiz) [inline detox/src/devices/android/ADB.js]: async ?

File: detox/src/artifacts/log/ios/SimulatorLogRecording.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  async doSave(artifactPath) {
+    await ensureMove(this._logPath, artifactPath, '.log');
+  }
+
+  async doDiscard() {
+    await fs.remove(this._logPath);
+  }
+
+  _createTail(file, prefix) {
+    const tail = new Tail(file, {
```
Comment (rotemmiz) [inline detox/src/artifacts/log/ios/SimulatorLogRecording.js]: I'm afraid tailing those files might be CPU intensive, let's think together if there's something out of process that we can do to avoid reading line by line.
Previously we have redirected std to files, and collected the files after each test. It's ugly, I agree, but much more efficient.

File: detox/src/artifacts/core/lifecycle/SnapshotterLifecycle.test.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+const SnapshotterLifecycle = require('./SnapshotterLifecycle');
+
+describe(SnapshotterLifecycle, () => {
+  it('should be requireable', () => {
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/artifacts/core/lifecycle/SnapshotterLifecycle.test.js]: These are stubs that do eslint role, sort of.

[the remark this answers is not in the document]
Comment (rotemmiz) [reply detox/src/artifacts/core/lifecycle/SnapshotterLifecycle.test.js]: Can you elaborate?

File: detox/src/artifacts/core/lifecycle/utils/ArtifactPathBuilder.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      return '';
+    }
+
+    const testDirectoryPrefix = this._buildTestDirectoryPrefix(testSummary);
+    const testArtifactsDirname = constructSafeFilename(testDirectoryPrefix, testSummary.fullName);
+
+    return testArtifactsDirname;
+  }
+
+  _assertConstructedPathIsStillInsideArtifactsRootDir(artifactPath, testSummary) {
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/artifacts/core/lifecycle/utils/ArtifactPathBuilder.js]: There is a fair reason. I’ll explain tomorrow.

File: detox/src/artifacts/core/lifecycle/utils/ArtifactPathBuilder.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+  _getStatusSign(testSummary) {
+    switch (testSummary.status) {
+      case 'passed': return '✓ ';
+      case 'failed': return '✗ ';
+      default: return '';
+    }
+  }
+
+  _getTestIndex(testSummary) {
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/artifacts/core/lifecycle/utils/ArtifactPathBuilder.js]: Done

File: detox/src/artifacts/core/lifecycle/utils/ArtifactPathBuilder.test.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+  describe('edge cases', () => {
+    beforeEach(() => {
+      strategy = new ArtifactPathBuilder({
+        artifactsRootDir: '/tmp',
+        getUniqueSubdirectory: _.constant('subdir'),
+      });
+    });
+
+    it('should defend against accidental resolving outside of root directory', () => {
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/artifacts/core/lifecycle/utils/ArtifactPathBuilder.test.js]: Yes. We allow. Or at least should. This test addresses totally different problem.

File: detox/src/devices/android/ADB.js (hunk 2 above)
Comment (noomorph) [reply detox/src/devices/android/ADB.js]: Yep. Missed. Thanks.

File: detox/src/utils/isChildProcessError.js [hunk 10]
```diff
@@ -0,0 +1,5 @@
+function isChildProcessError(error) {
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/utils/isChildProcessError.js]: No. I’ll decide a bit later if we need it in the bulletproof phase. If not, I’ll remove it.

File: detox/src/artifacts/ArtifactsManager.js [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    const videoRecorder = video({ artifactsRegistry });
+
+    this._hooks = _.compact([
+      this._createLogRecorderLifecycle(logRecorder),
+      this._createScreenshotterLifecycle(screenshotter),
+      this._createVideoRecorderLifecycle(videoRecorder),
+    ]);
+  }
+
+  _createLogRecorderLifecycle(logRecorder) {
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/artifacts/ArtifactsManager.js]: It’s not that easy. 2 are using RecorderLifecycle (different params: startup video is not recorded, opposed to the startup log) and 1 is using SnapshotterLifecycle. Merging all lifecycle into one heavy and parameter-rich is a doubtful endeavor, to my mind. Might complicate understanding of the implementation.

File: detox/src/artifacts/core/artifact/RecordingArtifact.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+  async doStart() {}
+
+  async doStop() {}
+
+  async doSave(artifactPath) {}
+
+  async doDiscard() {}
+
+  _assertRecordingIsNotBeingResumed() {
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/artifacts/core/artifact/RecordingArtifact.js]: You’re right. Now it is not user-facing, but honestly I was thinking that we wanted to make it available in next versions.

I think we indeed can remove most of the checks, but on other hand they served me a great debugging value whilst writing lifecycles. If anybody goes to change something there or adds a new lifecycle, it can help while testing.

Secondly, sometimes we might need a smarter logic inside. E.g., if on `SIGINT` we have to clean up allocated temporary artifact files, it is easier to call `.discard()` from ArtifactsManager for each artifact rather than implement status checks to see if we need to `.stop()` it before `.discard()`-ing it. Smarter implementation of `.discard()` gives less headache when you just need to `.discard()` something.

File: detox/src/artifacts/core/lifecycle/SnapshotterLifecycle.test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/artifacts/core/lifecycle/SnapshotterLifecycle.test.js]: I’ll turn them into normal tests within a couple of days

File: detox/src/artifacts/core/artifact/RecordingArtifact.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (rotemmiz) [reply detox/src/artifacts/core/artifact/RecordingArtifact.js]: I though that lifecycle is controlled by ArtifactManager, its a one time implementation, am I wrong ?

[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/artifacts/core/artifact/RecordingArtifact.js]: It is, but indirectly.
I prefer to keep lifecycles of a `SnapshotArtifact` and a `RecordingArtifact` as separate entities.
They are different enough in my opinion to not try to stitch them together.
In any case, `ArtifactsManager` is responsible for overall hooks (`onStart`, `onExit`, ...) execution and it has a couple of global (cross-lifecycle) handlers:
* It keeps registering artifacts, so it can discard them in an emergency.
* It provides `enqueueFinalizationTask` method to lifecycles - to coordinate finalization, e.g. to make it less CPU-intensive.

In other words, `ArtifactsManager` alone has enough of a headache to handle, so it delegates minutiae to "smaller" and more specific lifecycles per artifact type.

File: detox/src/artifacts/v2/videoRecorders/NonPausableVideoRecording.js [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    return this._onStop;
+  }
+
+  save() {
+    this._assertRecordingIsNotBeingDiscarded();
+
+    if (!this._onSave) {
+      this._onSave = this.stop().then(() => this._recording.save());
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/artifacts/v2/videoRecorders/NonPausableVideoRecording.js]: I need fine-grained control over the order of execution here, so I prefer not to use async-await here. Though, if I add enough tests that precisely verify that workflow works the way I expect it to, I may change the implementation to async-await later.

File: detox/src/utils/constructSafeFilename.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
+const DetoxRuntimeError = require('../errors/DetoxRuntimeError');
+const MAX_FILE_LENGTH = 255;
+
+/*
+  Trim filename to match filesystem limits (usually, not longer than 255 chars)
+ */
+function constructSafeFilename(prefix = '', trimmable = '', suffix = '') {
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/utils/constructSafeFilename.js]: There are no use cases with suffixes at the moment, to be honest. I just thought it would make more sense to write a signature like `(prefix, trimmable, suffix)` so it tells about the intent of function which is basically `return prefix + trimmable + suffix` or in a worse case `return prefix + trimmable.slice(-maxLen) + suffix`.

Maybe the source of your convolution is a multimethod implementation which treats a single parameter as a second parameter? I have no objections to simplify it, so it will look like:

```javascript
const testArtifactsDirname = constructSafeFilename(testIndexPrefix, testSummary.fullName);
const artifactFilename = constructSafeFilename('', artifactName);
```

[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/utils/constructSafeFilename.js]: Removed multimethod like I said before. Now signature should look less sophisticated, hopefully.

**Comment by tachtevrenidis** [comment]
screenshot support! yeah! :-)

**Comment by jwaldrip** [comment]
I could not be more excited about this feature!!!!

**Comment by haswalt** [comment]
Looking forward to this, last piece to remove fastlane and ruby from my build chain!

**Comment by rotemmiz** [comment]
🎉

**Comment by luisnaranjo733** [comment]
This is awesome. I think this will bring a ton of value to detox and enable QA scenarios that were previously not possible.

I'm very interested in using detox specifically for this functionality.
Judging by when this was merged and what I'm seeing in npmjs.com, I assume this will be available for the first time in 8.0.0-alpha.2.

Is there an ETA on that? Also, do you know when you are expecting to ship a stable 8.0.0 release?

**Comment by juddey** [comment]
This is fantastic. Thank you. ❤️ 

https://media.giphy.com/media/woZVO2B8ZQ55DqKsmN/giphy.gif

**Comment by mCodex** [comment]
Awesome work, thank you!
